# Rossmann Sales Data Cleaning
**Author:** Priyanshu Singh | Data Analyst  
**Dataset:** Rossmann Store Sales (Kaggle)  
**Tools:** Python, Pandas, PostgreSQL

## Objective
Clean and prepare 1,017,209 Rossmann retail sales records 
from 1,115 German stores for analysis and forecasting.

## Steps Covered
1. Load and explore raw data
2. Fix data types and extract date features
3. Remove invalid records using business logic
4. Clean store metadata and handle missing values
5. Merge datasets and validate
6. Load clean data to PostgreSQL

In [1]:
import pandas as pd

## Step 1 — Load Raw Data
Load train.csv (sales records) and store.csv (store metadata).
train.csv contains 1,017,209 daily sales records across 1,115 stores.
store.csv contains metadata for each store including type, assortment, and competition details.


In [2]:
df_train = pd.read_csv('../data/raw/train.csv', low_memory = False)
df_store = pd.read_csv('../data/raw/store.csv')

In [3]:
print(df_train.shape)

(1017209, 9)


In [4]:
df_train.head()

,Store,DayOfWeek,Date,Sales,Customers,Open,Promo,StateHoliday,SchoolHoliday
0,1,5,2015-07-31,5263,555,1,1,0,1
1,2,5,2015-07-31,6064,625,1,1,0,1
2,3,5,2015-07-31,8314,821,1,1,0,1
3,4,5,2015-07-31,13995,1498,1,1,0,1
4,5,5,2015-07-31,4822,559,1,1,0,1


In [5]:
print(df_train.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1017209 entries, 0 to 1017208
Data columns (total 9 columns):
 #   Column         Non-Null Count    Dtype 
---  ------         --------------    ----- 
 0   Store          1017209 non-null  int64 
 1   DayOfWeek      1017209 non-null  int64 
 2   Date           1017209 non-null  object
 3   Sales          1017209 non-null  int64 
 4   Customers      1017209 non-null  int64 
 5   Open           1017209 non-null  int64 
 6   Promo          1017209 non-null  int64 
 7   StateHoliday   1017209 non-null  object
 8   SchoolHoliday  1017209 non-null  int64 
dtypes: int64(7), object(2)
memory usage: 69.8+ MB
None


## Step 2 — Fix Data Types
Date column is stored as object (text). 
Converting to datetime enables extraction of Year, Month, Day, WeekOfYear features
which are critical for sales forecasting models.

In [6]:
df_train['Date'] = pd.to_datetime(df_train['Date'])

In [7]:
df_train["Date"].dtype

dtype('<M8[ns]')

In [8]:
df_train['Date'].sample(10)

354292   2014-08-27
172836   2015-02-26
489019   2014-04-19
419556   2014-06-21
444336   2014-05-29
980651   2013-02-02
447582   2014-05-26
975092   2013-02-07
767317   2013-08-13
673662   2013-11-05
Name: Date, dtype: datetime64[ns]

In [9]:
df_train['Year'] = df_train['Date'].dt.year
df_train['Month'] = df_train['Date'].dt.month
df_train['WeekOfYear'] = df_train['Date'].dt.isocalendar().week
df_train['Day'] = df_train['Date'].dt.day

In [10]:
df_train.head()

,Store,DayOfWeek,Date,Sales,Customers,Open,Promo,StateHoliday,SchoolHoliday,Year,Month,WeekOfYear,Day
0,1,5,2015-07-31,5263,555,1,1,0,1,2015,7,31,31
1,2,5,2015-07-31,6064,625,1,1,0,1,2015,7,31,31
2,3,5,2015-07-31,8314,821,1,1,0,1,2015,7,31,31
3,4,5,2015-07-31,13995,1498,1,1,0,1,2015,7,31,31
4,5,5,2015-07-31,4822,559,1,1,0,1,2015,7,31,31


In [11]:
print(df_train.isnull().sum())

Store            0
DayOfWeek        0
Date             0
Sales            0
Customers        0
Open             0
Promo            0
StateHoliday     0
SchoolHoliday    0
Year             0
Month            0
WeekOfYear       0
Day              0
dtype: int64


In [12]:
print(df_train.describe())

              Store     DayOfWeek                           Date  \
count  1.017209e+06  1.017209e+06                        1017209   
mean   5.584297e+02  3.998341e+00  2014-04-11 01:30:42.846061824   
min    1.000000e+00  1.000000e+00            2013-01-01 00:00:00   
25%    2.800000e+02  2.000000e+00            2013-08-17 00:00:00   
50%    5.580000e+02  4.000000e+00            2014-04-02 00:00:00   
75%    8.380000e+02  6.000000e+00            2014-12-12 00:00:00   
max    1.115000e+03  7.000000e+00            2015-07-31 00:00:00   
std    3.219087e+02  1.997391e+00                            NaN   

              Sales     Customers          Open         Promo  SchoolHoliday  \
count  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06   1.017209e+06   
mean   5.773819e+03  6.331459e+02  8.301067e-01  3.815145e-01   1.786467e-01   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00   
25%    3.727000e+03  4.050000e+02  1.000000e+00  0.000000e+00   0.0

In [13]:
df_train[df_train['Open'] == 0].shape

(172817, 13)

In [14]:
df_train[df_train['Sales'] ==0].shape

(172871, 13)

## Step 3 — Remove Invalid Records
**Business Logic Applied:**
- Dropped 172,817 rows where Open = 0 (closed stores have no sales — useless for forecasting)
- Dropped 54 rows where Open = 1 but Sales = 0 (store open but zero sales — data quality issue)

These decisions are based on domain understanding of retail operations.

In [15]:
df_train = df_train[df_train['Open'] != 0]
print(df_train.shape)

df_train = df_train[~((df_train['Open'] == 1) & (df_train['Sales'] == 0))]
print(df_train.shape)

(844392, 13)
(844338, 13)


In [16]:
df_train = df_train[df_train['Open'] != 0] 
print(df_train.shape)

(844338, 13)


In [17]:
df_train = df_train[~((df_train['Open'] ==1) & (df_train['Sales'] ==0))]
print(df_train.shape)

(844338, 13)


In [18]:
print(df_train.duplicated().sum())

0


In [19]:
print(df_train['Open'].unique())
print(df_train['Promo'].unique())

[1]
[1 0]


In [20]:
df_train.columns = df_train.columns.str.lower()

In [21]:
df_train.columns

Index(['store', 'dayofweek', 'date', 'sales', 'customers', 'open', 'promo',
       'stateholiday', 'schoolholiday', 'year', 'month', 'weekofyear', 'day'],
      dtype='object')

In [22]:
print(df_train.isnull().sum())
print(df_train.shape)
print(df_train.dtypes)

store            0
dayofweek        0
date             0
sales            0
customers        0
open             0
promo            0
stateholiday     0
schoolholiday    0
year             0
month            0
weekofyear       0
day              0
dtype: int64
(844338, 13)
store                     int64
dayofweek                 int64
date             datetime64[ns]
sales                     int64
customers                 int64
open                      int64
promo                     int64
stateholiday             object
schoolholiday             int64
year                      int32
month                     int32
weekofyear               UInt32
day                       int32
dtype: object


In [23]:
df_train.to_csv('../data/clean/train_clean.csv', index=False)

In [24]:
df_store.head()

,Store,StoreType,Assortment,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear,PromoInterval
0,1,c,a,1270.0,9.0,2008.0,0,NaN,NaN,NaN
1,2,a,a,570.0,11.0,2007.0,1,13.0,2010.0,"Jan,Apr,Jul,Oct"
2,3,a,a,14130.0,12.0,2006.0,1,14.0,2011.0,"Jan,Apr,Jul,Oct"
3,4,c,c,620.0,9.0,2009.0,0,NaN,NaN,NaN
4,5,a,a,29910.0,4.0,2015.0,0,NaN,NaN,NaN


In [25]:
df_store.describe()

,Store,CompetitionDistance,CompetitionOpenSinceMonth,CompetitionOpenSinceYear,Promo2,Promo2SinceWeek,Promo2SinceYear
count,1115.00000,1112.000000,761.000000,761.000000,1115.000000,571.000000,571.000000
mean,558.00000,5404.901079,7.224704,2008.668857,0.512108,23.595447,2011.763573
std,322.01708,7663.174720,3.212348,6.195983,0.500078,14.141984,1.674935
min,1.00000,20.000000,1.000000,1900.000000,0.000000,1.000000,2009.000000
25%,279.50000,717.500000,4.000000,2006.000000,0.000000,13.000000,2011.000000
50%,558.00000,2325.000000,8.000000,2010.000000,1.000000,22.000000,2012.000000
75%,836.50000,6882.500000,10.000000,2013.000000,1.000000,37.000000,2013.000000
max,1115.00000,75860.000000,12.000000,2015.000000,1.000000,50.000000,2015.000000


In [26]:
df_store.isnull().sum()

Store                          0
StoreType                      0
Assortment                     0
CompetitionDistance            3
CompetitionOpenSinceMonth    354
CompetitionOpenSinceYear     354
Promo2                         0
Promo2SinceWeek              544
Promo2SinceYear              544
PromoInterval                544
dtype: int64

In [27]:
df_store.columns = df_store.columns.str.lower()

## Step 4 — Clean Store Metadata
store.csv has missing values in competition and promotion columns.

**Filling Strategy:**
- CompetitionDistance → Median (outlier exists: max 75,860m vs median 2,325m)
- CompetitionOpenSinceMonth → Median (month is whole number 1-12)
- CompetitionOpenSinceYear → Median (year should be whole number)
- Promo2SinceWeek, Promo2SinceYear, PromoInterval → 0 (intentional nulls — stores with Promo2=0 never ran promotions)

In [28]:
median_column = ['competitiondistance', 'competitionopensincemonth', 'competitionopensinceyear']
for col in median_column:
    df_store[col] = df_store[col].fillna(df_store[col].median())

In [29]:
df_store.columns

Index(['store', 'storetype', 'assortment', 'competitiondistance',
       'competitionopensincemonth', 'competitionopensinceyear', 'promo2',
       'promo2sinceweek', 'promo2sinceyear', 'promointerval'],
      dtype='object')

In [30]:
zero_columns = ['promo2sinceweek', 'promo2sinceyear', 'promointerval']
for col in zero_columns:
    df_store[col] = df_store[col].fillna(0)

In [31]:
print(df_store.isnull().sum())

store                        0
storetype                    0
assortment                   0
competitiondistance          0
competitionopensincemonth    0
competitionopensinceyear     0
promo2                       0
promo2sinceweek              0
promo2sinceyear              0
promointerval                0
dtype: int64


In [32]:
df_store.duplicated().sum()

np.int64(0)

In [33]:
print(df_store.isnull().sum())
print(df_store.shape)
print(df_store.dtypes)

store                        0
storetype                    0
assortment                   0
competitiondistance          0
competitionopensincemonth    0
competitionopensinceyear     0
promo2                       0
promo2sinceweek              0
promo2sinceyear              0
promointerval                0
dtype: int64
(1115, 10)
store                          int64
storetype                     object
assortment                    object
competitiondistance          float64
competitionopensincemonth    float64
competitionopensinceyear     float64
promo2                         int64
promo2sinceweek              float64
promo2sinceyear              float64
promointerval                 object
dtype: object


In [34]:
type_conv = ['competitionopensincemonth', 'competitionopensinceyear', 'promo2sinceweek', 'promo2sinceyear']
for col in type_conv:
    df_store[col] = df_store[col].astype(int)

print(df_store.dtypes)

store                          int64
storetype                     object
assortment                    object
competitiondistance          float64
competitionopensincemonth      int64
competitionopensinceyear       int64
promo2                         int64
promo2sinceweek                int64
promo2sinceyear                int64
promointerval                 object
dtype: object


In [35]:
df_store.to_csv('../data/clean/store_clean.csv', index=False)

## Step 5 — Merge Datasets
Combining train.csv and store.csv on Store column using LEFT JOIN.
This adds store metadata (type, competition, promotions) to each sales record
enabling deeper business analysis.

In [36]:
df = pd.merge(df_train,df_store,on ='store',how='left')

In [37]:
print(df.shape)
df.head()

(844338, 22)


,store,dayofweek,date,sales,customers,open,promo,stateholiday,schoolholiday,year,...,day,storetype,assortment,competitiondistance,competitionopensincemonth,competitionopensinceyear,promo2,promo2sinceweek,promo2sinceyear,promointerval
0,1,5,2015-07-31,5263,555,1,1,0,1,2015,...,31,c,a,1270.0,9,2008,0,0,0,0
1,2,5,2015-07-31,6064,625,1,1,0,1,2015,...,31,a,a,570.0,11,2007,1,13,2010,"Jan,Apr,Jul,Oct"
2,3,5,2015-07-31,8314,821,1,1,0,1,2015,...,31,a,a,14130.0,12,2006,1,14,2011,"Jan,Apr,Jul,Oct"
3,4,5,2015-07-31,13995,1498,1,1,0,1,2015,...,31,c,c,620.0,9,2009,0,0,0,0
4,5,5,2015-07-31,4822,559,1,1,0,1,2015,...,31,a,a,29910.0,4,2015,0,0,0,0


In [38]:
df.isnull().sum()

store                        0
dayofweek                    0
date                         0
sales                        0
customers                    0
open                         0
promo                        0
stateholiday                 0
schoolholiday                0
year                         0
month                        0
weekofyear                   0
day                          0
storetype                    0
assortment                   0
competitiondistance          0
competitionopensincemonth    0
competitionopensinceyear     0
promo2                       0
promo2sinceweek              0
promo2sinceyear              0
promointerval                0
dtype: int64

In [39]:
df.to_csv('../data/clean/merged_cleaned.csv',index =False)

## Step 6 — Load to PostgreSQL
Loading clean merged data to PostgreSQL for SQL analysis.
PostgreSQL is used instead of MySQL as it is the standard database in European tech companies.

In [4]:
from dotenv import load_dotenv
import os
from sqlalchemy import create_engine
import pandas as pd

load_dotenv()
password = os.getenv('DB_PASSWORD')
engine = create_engine(f'postgresql://postgres:{password}@localhost:5432/rossmann')

df = pd.read_csv('../data/clean/merged_cleaned.csv', low_memory=False)
df_store = pd.read_csv('../data/clean/store_clean.csv')

df.to_sql('rossmann_sales', engine, if_exists='replace', index=False)
df_store.to_sql('rossmann_store', engine, if_exists='replace', index=False)

print("Data loaded successfully")

Data loaded successfully
